src/Core/Auth/UserFeatures/TwoFactorAuth/TwoFactorIsEnabledQuery.cs

```diff
@@ -71,13 +106,43 @@ public async Task<bool> TwoFactorIsEnabledAsync(ITwoFactorProvidersUser user)
             return false;
         }
 
-        return await TwoFactorEnabledAsync(
-                        user.GetTwoFactorProviders(),
-                        async () =>
-                        {
-                            var calcUser = await _userRepository.GetCalculatedPremiumAsync(userId.Value);
-                            return calcUser?.HasPremiumAccess ?? false;
-                        });
+        if (_featureService.IsEnabled(FeatureFlagKeys.PremiumAccessCacheCheck))
+        {
+            // Try to get premium status without fetching User entity if possible
+            bool hasPersonalPremium;
+            if (user is User userEntity)
+            {
+                hasPersonalPremium = userEntity.Premium;
+            }
+            else if (user is OrganizationUserUserDetails orgUserDetails)
+            {
+                hasPersonalPremium = orgUserDetails.Premium.GetValueOrDefault(false);
+            }
+            else
+            {
+                // Fallback: fetch the User entity
+                var fetchedUser = await _userRepository.GetByIdAsync(userId.Value);
+                if (fetchedUser == null)
+                {
+                    return false;
+                }
+                hasPersonalPremium = fetchedUser.Premium;
+            }
```

Comment:
This is trying to do the job of your new query, which is meant to be the only implementation of this logic. Some options:
- maybe Auth were wrong to remove the `GetPremium` method from `ITwoFactorProvidersUser`? It was unused, but it seems like you have a use for it here. Although if you go this way, I would put it in a separate `IPremiumUser` interface which `ITwoFactorProvidersUser` implements; then `IPremiumAccessQuery` accepts an `IPremiumUser` instead of a `User`.
- require the caller to pass in a `User` to begin with
- just always fetch the `User`, don't worry about it